app/controllers/katello/api/v2/repository_sets_controller.rb

```diff
@@ -125,25 +129,31 @@ def find_product_content
       if @product.present?
         @product_content = @product.product_content_by_id(params[:id])
       else
-        content = Katello::Content.find_by(:cp_content_id => params[:id], :organization_id => @organization[:id])
-        @product_content = Katello::ProductContent.find_by(:content_id => content.id)
+        content = Katello::Content.where(cp_content_id: params[:id], organization: @organization)
+        authorized_product_contents = Katello::ProductContent.joins(:product).merge(@product_scope)
+        @product_content = authorized_product_contents.joins(:content).merge(content).first
+        @product = @product_content&.product
       end
-      fail HttpErrors::NotFound, _("Couldn't find repository set with id '%s'.") % params[:id] if @product_content.nil?
-      @product = @product_content.product if @product.nil?
+      throw_resource_not_found(name: 'repository set', id: params[:id]) if @product_content.nil?
     end
 
-    def find_product_or_organization
+    def find_product
       if params[:product_id]
-        find_product
-      else
-        @organization = find_organization
+        @product = @product_scope.find_by(id: params[:product_id])
+        throw_resource_not_found(name: 'product', id: params[:product_id]) if @product.nil?
       end
     end
 
-    def find_product
-      @product = Product.find_by(:id => params[:product_id])
-      fail HttpErrors::NotFound, _("Couldn't find product with id '%s'") % params[:product_id] if @product.nil?
-      @organization = @product.organization
+    def set_readable_product_scope
+      @product_scope = Katello::Product.readable_by_subscription
```

Comment:
We could also do Katello::Subscription.readable.products I think. Maybe that's better so we can remove the new finders on product?